cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/provisioner/TetheringProvisioner.java

```diff
@@ -93,14 +93,14 @@ public Cluster createCluster(ProvisionerContext context) {
   }
 
   @Override
-  public ClusterStatus getClusterStatus(ProvisionerContext context, Cluster cluster) {
+  public Cluster getClusterStatus(ProvisionerContext context, Cluster cluster) {
     ClusterStatus status = cluster.getStatus();
-    return status == ClusterStatus.DELETING ? ClusterStatus.NOT_EXISTS : status;
+    return new Cluster(cluster, status == ClusterStatus.DELETING ? ClusterStatus.NOT_EXISTS : status);
```

Comment:
It looks like dup of NativeProvisioner.

Reply:
This kind of state machine changes should not be going in a getter call, that is not really a getter, but a mutation. There should be a better way to do it

Reply:
The cluster object has ClusterStatus as final. So a new cluster object is returned with updated status. There is no state transition happening here.

Reply:
Well, you pass one object and receive a same kind of object with different state. It's state transition even if underlying beans is immutable

Reply:
getClusterDetails has the same contract. 
https://github.com/cdapio/cdap/blob/d760fe1e785e58631f3be9fa773fecd5e71d0717/cdap-runtime-spi/src/main/java/io/cdap/cdap/runtime/spi/provisioner/Provisioner.java#L99

getClusterStatus earlier returned only a status. I wanted to return the error message as well, hence changed to using Cluster.

I am seeing Cluster as just a holder for carrying multiple pieces of information about the Dataproc cluster.

This behavior was there even earlier. @albertshau 
 any thoughts? https://github.com/cdapio/cdap/blob/8825d406bd24d8b4256a9962d6d336178829e1f1/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/provision/NativeProvisioner.java#L84

Reply:
the contract of this method is to just return the status. This is currently used when polling for cluster creation/deletion status. By returning an entire Cluster, the provisioner would be allowed to change the other parts of the Cluster information, which doesn't make sense.

This SPI change is also backwards incompatible and can't be done anyway. What about adding an optional message to the ClusterStatus?

Reply:
Adding a message to ClusterStatus should also work. But I am worried that will break compatibility as well.  How will the calling code check if a message field is present?

Here is another thought. To pass the error message out of band using ThreadLocal. Normally this is not a good solution, but given that we can't change the interface, I think it is acceptable. I did a prototype. Please look at the latest commit https://github.com/cdapio/cdap/pull/14509/commits/5f53780839ff7ce2d25c622cbad60440bb599a1c

Reply:
adding an optional field or adding a method is backwards compatible. Provisioners built against 6.7.x and older will still work with the new SPI. Only provisioners built against 6.8.0+ will be able to use the new field/method. Changing a method signature or remove a method is incompatible.

Reply:
Added a default method to the interface as suggested by Vitalii. And overrode it only for DataprocProvisioner.